Nodebalancer resource unexpected type error

<!-- Please keep this line, it helps us automate tagging issues. [issue-type:bug-report] -->

### Terraform Version

I am in the process of upgrading Terraform from `0.12.30` to `0.13.7`, in an effort to get my repo up to `1.0`+. I have been unable to refresh, plan, or apply with 0.13 because of the bug described here, so the state remains one from 0.12.

```
Terraform v0.13.7
+ provider registry.terraform.io/-/acme v2.5.3
+ provider registry.terraform.io/-/aws v3.55.0
+ provider registry.terraform.io/-/dnsimple v0.6.0
+ provider registry.terraform.io/-/linode v1.20.1
+ provider registry.terraform.io/-/random v3.1.0
+ provider registry.terraform.io/-/tls v3.1.0
+ provider registry.terraform.io/dnsimple/dnsimple v0.6.0
+ provider registry.terraform.io/hashicorp/aws v3.55.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/hashicorp/tls v3.1.0
+ provider registry.terraform.io/linode/linode v1.20.1
+ provider registry.terraform.io/vancluever/acme v2.5.3
```

### Effected Resource(s)

From my best effort to troubleshoot this, I believe these NodeBalancers are the reason for this issue.
```
$ terraform state pull | jq -r '.resources[] | select(.type=="linode_nodebalancer")'
{                                                                     
  "module": "module.grafana",
  "mode": "managed",
  "type": "linode_nodebalancer",
  "name": "grafana",
  "provider": "provider[\"registry.terraform.io/linode/linode\"]",
  "instances": [                
    {               
      "schema_version": 0,      
      "attributes": {  
        "client_conn_throttle": 0,                                    
        "created": "2019-11-10T04:52:48Z",
        "hostname": "nb-23-92-23-19.newark.nodebalancer.linode.com",
        "id": "53293",    
        "ipv4": "23.92.23.19",
        "ipv6": "2600:3c03:1::175c:1713",
        "label": "grafana",                                           
        "region": "us-east",                                          
        "tags": [],   
        "transfer": {         
          "in": "92.27030754089355",     
          "out": "71.42497444152832",
          "total": "163.69528198242188"
        },         
        "updated": "2019-11-10T04:52:48Z"
      }                            
    }                                                                 
  ]                                                                   
}
{                                                                     
  "module": "module.loki",
  "mode": "managed",
  "type": "linode_nodebalancer",
  "name": "loki",
  "provider": "provider[\"registry.terraform.io/linode/linode\"]",
  "instances": [                                                                                                                             
    {                                                                                                                                        
      "schema_version": 0,                                                                                                                   
      "attributes": {
        "client_conn_throttle": 0,                                                                                                           
        "created": "2020-03-06T03:22:21Z",
        "hostname": "nb-23-92-23-152.newark.nodebalancer.linode.com",
        "id": "61311",
        "ipv4": "23.92.23.152",                                       
        "ipv6": "2600:3c03:1::175c:1798",
        "label": "loki",
        "region": "us-east",
        "tags": [],
        "transfer": {
          "in": "1834.6708755493164",
          "out": "4230.0090799331665",
          "total": "6064.679955482483"
        },
        "updated": "2020-03-06T03:22:21Z"
      }
    }
  ]
}
{
  "module": "module.prometheus",
  "mode": "managed",
  "type": "linode_nodebalancer",
  "name": "prometheus",
  "provider": "provider[\"registry.terraform.io/linode/linode\"]",
  "instances": [
    {
      "schema_version": 0,
      "attributes": {
        "client_conn_throttle": 0,
        "created": "2019-11-21T02:16:15Z",
        "hostname": "nb-23-92-23-10.newark.nodebalancer.linode.com",
        "id": "54187",
        "ipv4": "23.92.23.10",
        "ipv6": "2600:3c03:1::175c:170a",
        "label": "prometheus",
        "region": "us-east",
        "tags": [],
        "transfer": {
          "in": "86.8517951965332",
          "out": "14.579236030578613",
          "total": "101.43103122711182"
        },
        "updated": "2019-11-21T02:16:15Z"
      }
    }
  ]
}
```

### Terraform Configuration Files

https://git.stj.io/stjacobs/infra/src/master/terraform

### Debug Output

```
$ terraform refresh

# ... lots of states being refreshed

Error: missing expected [

Error: missing expected [

Error: missing expected [
```

When looking through the debug logs, the following stuck out, and is the reason I opened the issue here. There are 3 cases of `ERROR` (and 3 Nodebalancers) in the logs.

```
2021/08/21 11:47:55 [WARN] UpgradeResourceState: unexpected type cty.List(cty.Object(map[string]cty.Type{"in":cty.Number, "out":cty.Number, "
total":cty.Number})) for map in json state
2021/08/21 11:47:55 [TRACE] UpgradeResourceState: schema version of module.grafana.linode_nodebalancer.grafana is still 0; calling provider "
linode" for any other minor fixups
2021/08/21 11:47:55 [TRACE] GRPCProvider: UpgradeResourceState
2021-08-21T11:47:55.942-0400 [INFO]  plugin.terraform-provider-linode_v1.20.1: 2021/08/21 11:47:55 [WARN] unexpected type cty.List(cty.Object(map[string]cty.Type{"in":cty.Number, "out":cty.Number, "total":cty.Number})) for map in json state: timestamp=2021-08-21T11:47:55.941-0400
2021/08/21 11:47:55 [ERROR] eval: *terraform.EvalReadState, err: missing expected [
2021/08/21 11:47:55 [ERROR] eval: *terraform.EvalSequence, err: missing expected [
# ... more logs
2021/08/21 11:47:55 [WARN] UpgradeResourceState: unexpected type cty.List(cty.Object(map[string]cty.Type{"in":cty.Number, "out":cty.Number, "
total":cty.Number})) for map in json state
2021/08/21 11:47:55 [TRACE] UpgradeResourceState: schema version of module.loki.linode_nodebalancer.loki is still 0; calling provider "linode
" for any other minor fixups
2021/08/21 11:47:55 [TRACE] GRPCProvider: UpgradeResourceState
2021-08-21T11:47:55.980-0400 [INFO]  plugin.terraform-provider-linode_v1.20.1: 2021/08/21 11:47:55 [WARN] unexpected type cty.List(cty.Object
(map[string]cty.Type{"in":cty.Number, "out":cty.Number, "total":cty.Number})) for map in json state: timestamp=2021-08-21T11:47:55.980-0400
2021/08/21 11:47:55 [ERROR] eval: *terraform.EvalReadState, err: missing expected [
2021/08/21 11:47:55 [ERROR] eval: *terraform.EvalSequence, err: missing expected [
```

### Expected Behavior

State is refreshed. Eventually a plan is able to be created and applied, updating the state to be compatible with 0.13.

### Actual Behavior

Errors occur before state can be refreshed.

### Steps to Reproduce

On Terraform 0.13, in an attempt to refresh a state from 0.12 for the first time...

- I am able to refresh with terraform-provider-linode version 1.13.4
- I am not able to refresh with terraform-provider-linode version 1.14.0

I think you should be able to create a Nodebalancer with Terraform 0.12 and terraform-provider-linode 0.13.4 and then go to Terraform 0.13 and terraform-provider-linode 0.14.0 or greater.

Looking at the logs, I believe this is the change that made the difference.

https://github.com/linode/terraform-provider-linode/pull/203

I even tried applying with Terraform 0.13 and terraform-provider-linode 1.13.4, which was successful. Upgrading terraform-provider-linode to 1.20.1 while still on Terraform 0.13 was again unsuccessful.

--- a/linode/resource_linode_nodebalancer.go
+++ b/linode/resource_linode_nodebalancer.go
@@ -44,6 +44,14 @@ func resourceLinodeNodeBalancer() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
+		SchemaVersion: 1,
+		StateUpgraders: []schema.StateUpgrader{
+			{
+				Type:    resourceLinodeNodeBalancerV0().CoreConfigSchema().ImpliedType(),
+				Upgrade: resourceLinodeNodeBalancerV0Upgrade,
+				Version: 0,
+			},
+		},
 		Schema: map[string]*schema.Schema{
 			"label": {
 				Type:        schema.TypeString,
@@ -103,6 +111,53 @@ func resourceLinodeNodeBalancer() *schema.Resource {
 			},
 		},
 	}
+}
+
+func resourceLinodeNodeBalancerV0() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"transfer": {
+				Type:     schema.TypeMap,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceLinodeNodeBalancerV0Upgrade(ctx context.Context,
+	rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	oldTransfer, ok := rawState["transfer"].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("failed to upgrade state: transfer key does not exist")
+	}
+
+	newTransfer := []map[string]interface{}{
+		{
+			"in":    0.0,
+			"out":   0.0,
+			"total": 0.0,
+		},
+	}
+
+	for key, val := range oldTransfer {
+		val := val.(string)
+
+		// This is necessary because it is possible old versions of the state have empty transfer fields
+		// that must default to zero.
+		if val == "" {
+			continue
+		}
+
+		result, err := strconv.ParseFloat(val, 64)
+		if err != nil {
+			return nil, fmt.Errorf("failed to upgrade state: %v", err)
+		}
+
+		newTransfer[0][key] = result
+	}
+
+	rawState["transfer"] = newTransfer
+	return rawState, nil
 }
 
 func resourceLinodeNodeBalancerRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {

--- a/linode/resource_linode_nodebalancer_config.go
+++ b/linode/resource_linode_nodebalancer_config.go
@@ -40,6 +40,14 @@ func resourceLinodeNodeBalancerConfig() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: resourceLinodeNodeBalancerConfigImport,
 		},
+		SchemaVersion: 1,
+		StateUpgraders: []schema.StateUpgrader{
+			{
+				Type:    resourceLinodeNodeBalancerConfigV0().CoreConfigSchema().ImpliedType(),
+				Upgrade: resourceLinodeNodeBalancerConfigV0Upgrade,
+				Version: 0,
+			},
+		},
 		Schema: map[string]*schema.Schema{
 			"nodebalancer_id": {
 				Type:        schema.TypeInt,
@@ -188,6 +196,52 @@ func resourceLinodeNodeBalancerConfig() *schema.Resource {
 			},
 		},
 	}
+}
+
+func resourceLinodeNodeBalancerConfigV0() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"node_status": {
+				Type:     schema.TypeMap,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceLinodeNodeBalancerConfigV0Upgrade(ctx context.Context,
+	rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	oldTransfer, ok := rawState["node_status"].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("failed to upgrade state: node_status key does not exist")
+	}
+
+	newTransfer := []map[string]interface{}{
+		{
+			"down": 0,
+			"up":   0,
+		},
+	}
+
+	for key, val := range oldTransfer {
+		val := val.(string)
+
+		// This is necessary because it is possible old versions of the state have empty transfer fields
+		// that must default to zero.
+		if val == "" {
+			continue
+		}
+
+		result, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse state: %v", err)
+		}
+
+		newTransfer[0][key] = result
+	}
+
+	rawState["node_status"] = newTransfer
+	return rawState, nil
 }
 
 func resourceLinodeNodeBalancerConfigImport(

--- a/linode/resource_linode_nodebalancer_config_test.go
+++ b/linode/resource_linode_nodebalancer_config_test.go
@@ -3,6 +3,7 @@ package linode
 import (
 	"context"
 	"fmt"
+	"reflect"
 	"strconv"
 	"testing"
 
@@ -251,6 +252,64 @@ func TestAccLinodeNodeBalancerConfig_proxyProtocol(t *testing.T) {
 			},
 		},
 	})
+}
+
+func TestLinodeNodeBalancerConfig_UpgradeV0(t *testing.T) {
+	t.Parallel()
+
+	oldState := map[string]interface{}{
+		"node_status": map[string]interface{}{
+			"down": "13",
+			"up":   "37",
+		},
+	}
+
+	desiredState := map[string]interface{}{
+		"node_status": []map[string]interface{}{
+			{
+				"down": 13,
+				"up":   37,
+			},
+		},
+	}
+
+	newState, err := resourceLinodeNodeBalancerConfigV0Upgrade(context.Background(), oldState, nil)
+	if err != nil {
+		t.Fatalf("error migrating state: %v", err)
+	}
+
+	if !reflect.DeepEqual(desiredState, newState) {
+		t.Fatalf("expected %v, got %v", desiredState, newState)
+	}
+}
+
+func TestLinodeNodeBalancerConfig_UpgradeV0Empty(t *testing.T) {
+	t.Parallel()
+
+	oldState := map[string]interface{}{
+		"node_status": map[string]interface{}{
+			"down": "",
+			"up":   "",
+		},
+	}
+
+	desiredState := map[string]interface{}{
+		"node_status": []map[string]interface{}{
+			{
+				"down": 0,
+				"up":   0,
+			},
+		},
+	}
+
+	newState, err := resourceLinodeNodeBalancerConfigV0Upgrade(context.Background(), oldState, nil)
+	if err != nil {
+		t.Fatalf("error migrating state: %v", err)
+	}
+
+	if !reflect.DeepEqual(desiredState, newState) {
+		t.Fatalf("expected %v, got %v", desiredState, newState)
+	}
 }
 
 func testAccCheckLinodeNodeBalancerConfigExists(s *terraform.State) error {

--- a/linode/resource_linode_nodebalancer_test.go
+++ b/linode/resource_linode_nodebalancer_test.go
@@ -3,6 +3,7 @@ package linode
 import (
 	"context"
 	"fmt"
+	"reflect"
 	"strconv"
 	"testing"
 
@@ -114,6 +115,68 @@ func TestAccLinodeNodeBalancer_update(t *testing.T) {
 			},
 		},
 	})
+}
+
+func TestLinodeNodeBalancer_UpgradeV0(t *testing.T) {
+	t.Parallel()
+
+	oldState := map[string]interface{}{
+		"transfer": map[string]interface{}{
+			"in":    "1337",
+			"out":   "1338",
+			"total": "1339",
+		},
+	}
+
+	desiredState := map[string]interface{}{
+		"transfer": []map[string]interface{}{
+			{
+				"in":    1337.0,
+				"out":   1338.0,
+				"total": 1339.0,
+			},
+		},
+	}
+
+	newState, err := resourceLinodeNodeBalancerV0Upgrade(context.Background(), oldState, nil)
+	if err != nil {
+		t.Fatalf("error migrating state: %v", err)
+	}
+
+	if !reflect.DeepEqual(desiredState, newState) {
+		t.Fatalf("expected %v, got %v", desiredState, newState)
+	}
+}
+
+func TestLinodeNodeBalancer_UpgradeV0Empty(t *testing.T) {
+	t.Parallel()
+
+	oldState := map[string]interface{}{
+		"transfer": map[string]interface{}{
+			"in":    "",
+			"out":   "",
+			"total": "",
+		},
+	}
+
+	desiredState := map[string]interface{}{
+		"transfer": []map[string]interface{}{
+			{
+				"in":    0.0,
+				"out":   0.0,
+				"total": 0.0,
+			},
+		},
+	}
+
+	newState, err := resourceLinodeNodeBalancerV0Upgrade(context.Background(), oldState, nil)
+	if err != nil {
+		t.Fatalf("error migrating state: %v", err)
+	}
+
+	if !reflect.DeepEqual(desiredState, newState) {
+		t.Fatalf("expected %v, got %v", desiredState, newState)
+	}
 }
 
 func testAccCheckLinodeNodeBalancerExists(s *terraform.State) error {

